msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/IMDSManagedIdentity.java

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class IMDSManagedIdentity extends AbstractManagedIdentitySource{
+
+    // IMDS constants. Docs for IMDS are available here https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http
+    private final static Logger LOG = LoggerFactory.getLogger(IMDSManagedIdentity.class);
+    private static URI DEFAULT_IMDS_ENDPOINT;
+
+    static {
+        try {
+            DEFAULT_IMDS_ENDPOINT = new URI("http://169.254.169.254/metadata/identity/oauth2/token");
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String imdsTokenPath = "/metadata/identity/oauth2/token";
+    private String imdsApiVersion = "2018-02-01";
+    private static String defaultMessage = "[Managed Identity] Service request failed.";
+
+    private static String identityUnavailableError = "[Managed Identity] Authentication unavailable. The requested identity has not been assigned to this resource.";
+    private static String gatewayError = "[Managed Identity] Authentication unavailable. The request failed due to a gateway error.";
+
+    private URI imdsEndpoint;
+
+    public IMDSManagedIdentity(MsalRequest msalRequest,
+                               ServiceBundle serviceBundle) {
+        super(msalRequest, serviceBundle, ManagedIdentitySourceType.Imds);
+        ManagedIdentityParameters parameters = (ManagedIdentityParameters) msalRequest.requestContext().apiParameters();
+        IEnvironmentVariables environmentVariables = ((ManagedIdentityParameters) msalRequest.requestContext().apiParameters()).environmentVariables == null ?
+                new EnvironmentVariables() :
+                parameters.environmentVariables;
+        if (!StringHelper.isNullOrBlank(environmentVariables.getEnvironmentVariable(IEnvironmentVariables.AZURE_POD_IDENTITY_AUTHORITY_HOST))){
+            LOG.info("[Managed Identity] Environment variable AZURE_POD_IDENTITY_AUTHORITY_HOST for IMDS returned endpoint: " + environmentVariables.getEnvironmentVariable(IEnvironmentVariables.AZURE_POD_IDENTITY_AUTHORITY_HOST));
+            try {
+                imdsEndpoint = new URI(environmentVariables.getEnvironmentVariable(IEnvironmentVariables.AZURE_POD_IDENTITY_AUTHORITY_HOST));
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
```

Comment:
Just for my understanding - we're going with RuntimeException here to keep the code clean? We never expect `URISyntaxException` to occur, so we don't to have it as a checked exception in the public API?

Reply:
This should be MsalManagedIdentityException. Ideally this will never happen. Updating here and rest of the places.